geoutils/geoviewer.py

```diff
@@ -1,98 +1,170 @@
+        type=int,
+        default=2000,
+        help="int, image size is limited to max_size**2 for faster reading/displaying (Default is 2000).",
+    )
+    parser.add_argument(
+        "-save",
+        dest="save",
+        type=str,
+        default="",
+        help="str, filename to the output filename to save to disk (Default is displayed on screen).",
+    )
+    parser.add_argument(
+        "-dpi",
+        dest="dpi",
+        type=str,
+        default="default",
+        help="int, dpi value to use when saving figure (Default is from rcParams).",
+    )
+    parser.add_argument(
+        "-nodata",
+        dest="nodata",
+        type=str,
+        default="default",
+        help="float, no data value (Default is read from file metadata).",
+    )
+    parser.add_argument(
+        "-noresampl",
+        dest="noresampl",
+        default=False,
+        action="store_true",
+        help="True or False, if False then allow dynamic image downscaling, if True, prevent it.",
+    )
 
     args = parser.parse_args()
 
     return args
 
 
-def main():
+def main() -> None:
 
     # Parse arguments
     args = getparser()
 
-    ## Load image metadata ##
+    # Load image metadata #
     img = Raster(args.filename, load_data=False)
     xmin, xmax, ymin, ymax = img.bounds
 
-    ## Resample if image is too large ##
-    if ((img.width > args.max_size) or (img.height > args.max_size)) \
-       & (not args.noresampl):
-        dfact = max(
-            int(img.width/args.max_size),
-            int(img.height/args.max_size)
-        )
-        print("Image will be downsampled by a factor {}.".format(dfact))
-        new_shape = (img.count, int(img.height/dfact), int(img.width/dfact))
+    # Resample if image is too large #
+    if ((img.width > args.max_size) or (img.height > args.max_size)) & (not args.noresampl):
+        dfact = max(int(img.width / args.max_size), int(img.height / args.max_size))
+        print(f"Image will be downsampled by a factor {dfact}.")
+        new_shape: tuple[Any, ...] | None = (img.count, int(img.height / dfact), int(img.width / dfact))
     else:
         new_shape = None
 
-    ## Read image ##
+    # Read image #
     img.load(out_shape=new_shape)
 
     # Set no data value
-    if args.nodata == 'default':
+    if args.nodata == "default":
         nodata = img.nodata
     else:
         try:
             nodata = float(args.nodata)
         except ValueError:
-            raise ValueError("ERROR: nodata must be a float, currently set to %s" %args.nodata)
+            raise ValueError("ERROR: nodata must be a float, currently set to %s" % args.nodata)
 
         img.set_ndv(nodata)
 
-    ## Set default parameters ##
+    # Set default parameters #
 
     # vmin
     if args.vmin is not None:
         try:
-            vmin = float(args.vmin)
-        except ValueError:   # Case is not a number
-            perc, _ = args.vmin.split('%')
+            vmin: float | None = float(args.vmin)
```

Comment:
Will it be needed to define the type of ALL variables??

Reply:
No, it's because `vmin` may be either a float or a None, so when it's first defined, it needs to be typed.

The incorrect way:
```python
vmin = 3.0

vmin = None  # this is wrong because vmin is supposed to be a float
```
The correct way
```python
vmin: float | None = 3.0

vmin = None  # This is now fine because the type is Optional[float]
```

In another case when the type does not change, types are dynamically understood:
```python
float_value = 1.1

float_value  = 1.7
```

Reply:
Ok, but then we need to specify the type of any variable that may have several types...
Which of the pre-commit requires this?

Reply:
mypy

Imagine having a function that returns `Optional[float]` (read: it may return a NoneType). If a variable is already typed to `float`, we need to handle this potential NoneType and the error is therefore very useful.

Reply:
Yes, but vmin is just a local variable that is never used outside the function... Anyway, even if I don't quite get the point, that is not really the issue here.
The question is, if I write the same code without writing the type of vmin, what will happen? How will I know there's something wrong in the code and how to fix it?